Nextcloud.md updated & Added the tutorial for OIDC & Authelia

Normanras opened the pull request:
Added Tutorial for Nextcloud/OIDC/Authelia/LLDAP & Updated minor changes on Nextcloud.md

Normanras commented:
Forgot to run cargo... closing this PR and resubmitting.

Normanras commented:
no errors with cargo fmt

nitnelave commented on example_configs/nextcloud.md:
```diff
@@ -5,8 +5,23 @@
 * A technical user (ex: `ro_admin`), member of `lldap_strict_readonly` or `lldap_password_manager`
-* Several accounts, members of `users` group will be authorized to log in Nextcloud (eg neither `admin` nor `ro_admin`)
-* Some "application" groups, let's say `friends` and `family`: users in Nextcloud will be able to share files and view people in dynamic lists only to members of their own group(s)
+* A group called `users`. Note: This is _not_ the same as the `Users` designation in the LLDAP Web UI. This must be a whole new group that you've created and called `users`. 
```
How about we change the config here to name it `nextcloud_users`? It would be clearer with less risk for confusion.

nitnelave commented on example_configs/nextcloud.md:
```diff
@@ -5,8 +5,23 @@
-* Several accounts, members of `users` group will be authorized to log in Nextcloud (eg neither `admin` nor `ro_admin`)
-* Some "application" groups, let's say `friends` and `family`: users in Nextcloud will be able to share files and view people in dynamic lists only to members of their own group(s)
+* A group called `users`. Note: This is _not_ the same as the `Users` designation in the LLDAP Web UI. This must be a whole new group that you've created and called `users`. 
+* Members of `users` group will be authorized to log in Nextcloud (eg neither `admin` nor `ro_admin`, who are part of the `lldap_strict_readonly` or `lldap_password_manager` group.)
```
I'm not sure the "eg" comment part is super clear. We can probably just remove it.

Normanras replied:
👍🏼

nitnelave commented on example_configs/nextcloud.md:
```diff
@@ -5,8 +5,23 @@
+    3. `friends`
+    4. `lldap_strict_readonly`
+    5. `lldap_password_manager`
+    6. `admin`
```
`lldap_admin`

nitnelave commented on example_configs/nextcloud.md:
```diff
@@ -5,8 +5,23 @@
+    5. `lldap_password_manager`
+    6. `admin`
+* 1 admin user in one or both of the following groups:
+    1. `lldap_password_manager`
```
or `lldap_admin`

nitnelave commented on example_configs/nextcloud.md:
```diff
@@ -5,8 +5,23 @@
-* Some "application" groups, let's say `friends` and `family`: users in Nextcloud will be able to share files and view people in dynamic lists only to members of their own group(s)
+* A group called `users`. Note: This is _not_ the same as the `Users` designation in the LLDAP Web UI. This must be a whole new group that you've created and called `users`. 
+* Members of `users` group will be authorized to log in Nextcloud (eg neither `admin` nor `ro_admin`, who are part of the `lldap_strict_readonly` or `lldap_password_manager` group.)
+* Some "application" groups, let's say `friends` and `family`: users in Nextcloud will be able to share files and view people in dynamic lists only to members of their own group(s).  
```
I think we should clarify that users in `family` or `friends` should _also_ be in `nextcloud_users`

nitnelave commented on example_configs/nextcloud.md:
````diff
@@ -79,6 +95,8 @@ We want only users from the `users` group to be allowed to log in Nextcloud :
 (&(objectclass=person)(memberOf=cn=users,ou=groups,dc=example,dc=com))
 ```
 
+_Note:_ The `cn=users` section of the LDAP query is the group you setup at the begining of this tutorial! It is _not_ the users section as seen in the LLDAP web ui. If you're using different group names in LLDAP, be sure to update the `cn=` to the appropriate group name. 
````
Once we rename that to `nextcloud_users`, we can remove this line.

nitnelave commented on example_configs/nextcloud_oidc_authelia.md:
```diff
@@ -0,0 +1,81 @@
+
+Set up [LLDAP (Light Lightweight Directory Access Protocol](https://github.com/nitnelave/lldap) according to its standard documentation. 
+
+Add a group and note the name you use. For this tutorial, we're using the group `users`. Create a new user and add it to the `users` group. 
```
Maybe we should use `nextcloud_users` here as well? Since we're connecting with nextcloud.

Normanras replied:
Yes, I'm going to chance it across both docs. I'll also check if any other examples reference nextcloud and update there, but let me know if you know of any specifically.

nitnelave commented on example_configs/nextcloud_oidc_authelia.md:
```diff
@@ -0,0 +1,81 @@
+
+## LLDAP
+
+Set up [LLDAP (Light Lightweight Directory Access Protocol](https://github.com/nitnelave/lldap) according to its standard documentation. 
```
I think there's no need to spell out LLDAP :D

Normanras replied:
Oversharing from when I first wrote it as a gist :)

nitnelave commented on example_configs/nextcloud_oidc_authelia.md:
```diff
@@ -0,0 +1,81 @@
+Add a group and note the name you use. For this tutorial, we're using the group `users`. Create a new user and add it to the `users` group. 
+
+#### Optional:
+Once setup, add an admin or config user and add to the `lldap_strict_readonly` group. This will be the config account used for Nextcloud to read your groups and users from the server. 
```
`lldap_strict_readonly` or `lldap_password_manager` ? Can the app modify passwords?

Normanras replied:
Good point, I didn't explicitly note it in the nextcloud ldap settings and it comes unselected here. Especially because of the plain text transport, I'm a bit skeptical about choosing anything besides readonly. As the creator, what do you think?

![Screenshot 2022-11-14 at 2 33 55 PM](https://user-images.githubusercontent.com/44226464/201749648-6441a8a6-9bbf-406d-987c-8017d87ccc7c.png)

nitnelave commented on example_configs/nextcloud_oidc_authelia.md:
```diff
@@ -0,0 +1,81 @@
+* **Client ID:** nextcloud
+* **Client Secret:** nextcloud_client_secret
+
+Login to your Nextcloud instance as an admin. Setup an LDAP server connection according to [LLDAP's Documentation](https://github.com/nitnelave/lldap/blob/main/example_configs/nextcloud.md).Once your configuration is setup and Nextcloud doesn't show any errors, you can move onto the next step. It's also a good idea to test this by logging into your Nextcloud instance with your LLDAP credentials.  ...
```
Nit: space before "Once".

Normanras replied:
Nit is worth it! I appreciate this catch

nitnelave commented on example_configs/nextcloud_oidc_authelia.md:
```diff
@@ -0,0 +1,81 @@
+
+Next, navigate to the `Apps` section. 
+
+[![nextcloud_apps.png](images/nextcloud_apps.png)
```
Can you add the images?

nitnelave replied:
And you've got an extra "[" on this line.
Tip: On GitHub, on the PR screen, you can click the three dots in the upper right corner, then "View this file" to see the markdown rendered. You'll see the image missing, the formatting and so on.

Normanras replied:
I previewed and thought I got all the stray "[" but I see I missed some. Also, I updated the images and previewed the markdown locally, which pulled and loaded from that dir as expected. Are the images not being uploaded? I'm now thinking I Pr'd the wrong branch? Let me check over it again and I'll comment here or message you on discord.

nitnelave replied:
You can see the "Files changed" in the PR: there are only the two markdown files, no new images.

Normanras replied:
I think I know what I did wrong. I'll finish my edits and push changes. I bet I only pushed the two files.

nitnelave commented on example_configs/nextcloud_oidc_authelia.md:
```diff
@@ -0,0 +1,81 @@
+### Conclusion
+And that's it! Assuming all the settings that worked for me, work for you, you should be able to login using OpenID Connect via Authelia. If you find any errors, it's a good idea to keep a document of all your settings from Authelia/Nextcloud/LLDAP etc so that you can easily reference and ensure everything lines up.
+
+Please comment here if you face any issues and we can test out other settings. 
```
Maybe:
"If you have any issues, please create a `[discussion](https://github.com/nitnelave/lldap/discussions)` or join the `[Discord](https://discord.gg/h5PEdRMNyP)`."

nitnelave reviewed (changes requested):
Sorry for the many comments, but it's a rather complicated setup, so I want to try to eliminate as much ambiguity as possible.

Normanras commented:
> Sorry for the many comments, but it's a rather complicated setup, so I want to try to eliminate as much ambiguity as possible.

No problem at all! I would rather go through and get it right instead of having to go back to redo the work in a new PR. I'm not sure why there are so many backtick errors though, I made sure to preview everything locally and in my fork and it looked normal. Thanks for finding and commenting on them.

Normanras commented on example_configs/nextcloud.md:
```diff
@@ -5,8 +5,26 @@
+    5. `lldap_password_manager`
+    6. 'lldap_admin'
+    7. `admin`
+* 1 admin user in one or both of the following groups:
```
A weird example of the ticks. Please let me know when I commit the changes again if you see the tick problem again:
<img width="397" alt="Screenshot 2022-11-15 at 11 04 40 AM" src="https://user-images.githubusercontent.com/44226464/201969120-d86c5a86-7396-47f0-8ee6-e203a4b279ee.png">

Also, resolved, removed line 20 with the `lldap_admin`.

Normanras commented on example_configs/nextcloud.md:
````diff
@@ -80,7 +99,6 @@ We want only users from the `users` group to be allowed to log in Nextcloud :
 ```
 
 ![login configuration page](images/nextcloud_loginfilter.png)
````
it should really be retaken and added it. I've done that.

Normanras commented on example_configs/nextcloud.md:
```diff
@@ -100,7 +118,6 @@ You can use the menus for this part : select `groupOfUniqueNames` in the first m
-
 ## Sharing restrictions
 
 Go to Settings > Administration > Sharing and check following boxes :
```
Fair enough. I took a new screenshot too.

Normanras commented on example_configs/nextcloud_oidc_authelia.md:
```diff
@@ -0,0 +1,81 @@
+* **Application Root URL:** https://nextcloud.example.com
+* **Authelia Root URL:** https://auth.example.com
+* **Client ID:** nextcloud
+* **Client Secret:** nextcloud_client_secret
```
It comes from [Authelia](https://www.authelia.com/configuration/identity-providers/open-id-connect/#secret), you define it when you're setting up the integration. I can be a random string that is used to authenticate the two apps talking. I may be making too many assumptions and should add something to the top of this doc saying that this tutorial expects that you have Authelia up and running and understand how to access and update the config file, etc.

Normanras commented:
Okay, I feel like we're going down a path here, but for this PR I've implemented all the requested changes and added some extra links and references (such as Authelia's client secret/id). I also removed from over-explaining lines and added a section at the top detailing assumptions. 

I don't have time in the coming weeks, but I think revamping this with a bit more explanation with the Authelia setup might make this more clear. I'm open to your feedback, but let me know what you see in this commit. I'm mostly concerned that the backticks errors will persist, so I'll have to open it in another editor if it continues to revert to apostraphes.

nitnelave reviewed (approved):
It looks good! Can you just remove the example_configs/images/.DS_Store file?

Normanras commented on example_configs/nextcloud.md:
```diff
@@ -5,24 +5,22 @@
-    6. 'lldap_admin'
-    7. `admin`
-* 1 admin user in one or both of the following groups:
+    6. `admin`
```
oh my 🫣 you must hate this pr. hopefully the last error! In my original setup, I had an admin group (which is pointless since `lldap_admin exists`) and completely overlooked it when I moved from my gist to here. Thanks again.

Normanras commented:
> It looks good! Can you just remove the example_configs/images/.DS_Store file?

Yes, done.

nitnelave commented on example_configs/nextcloud_oidc_authelia.md:
```diff
@@ -0,0 +1,87 @@
+If you're here, there are some assumptions being made about access and capabilities you have on your system:
+1. You have Authelia up and running, understand its functionality, and have read through the documentation. 
+2. You have [LLDAP](https://github.com/nitnelave/lldap) up and running. 
+3. You have Nextcloud and LLDAP communicating and without any config errors. See the [example config for Nextcloud](example_configs/nextcloud.md)
```
The link for the Nextcloud config is relative, so it should be just `nextcloud.md`, not `example_configs/nextcloud.md`

nitnelave commented on example_configs/nextcloud_oidc_authelia.md:
```diff
@@ -0,0 +1,87 @@
+* Allow Users to Connect Social Logins with their Account
+* Prevent creating an account if the email address exists in another account
+* Update user profile every login
+![nextcloud_sociallogin_checkboxes](images/nextcloud_sociallogin_checkboxes.png)
```
There should be a blank line above the image.

Normanras replied:
How about after the images too? Line 69/70, for example.

nitnelave reviewed (commented):
Almost there! It should be the last round. Sorry I didn't catch these earlier.

nitnelave commented on example_configs/nextcloud_oidc_authelia.md:
```diff
@@ -0,0 +1,87 @@
+
+#### Some Notes
+* The *scope* should be the same as the scope that was setup in [Authelia's OpenID Integration](https://www.authelia.com/integration/openid-connect/nextcloud/#authelia). Here's an example from Authelia:
+![Authelia OpenID Config](images/authelia_openid_config.png)
```
I think you didn't commit it, I don't see the new line.

But with respect to whether the new line is needed or not, it's just a matter of alignment. The images should basically always be in a separate block, left-aligned. To achieve that, the easiest is to surround them with empty lines. You can double-check how it looks in the preview, if you use a wide screen (a small screen will just wrap it around and make it look like it's left-aligned).

Normanras replied:
👍🏼

nitnelave commented:
Thanks a lot!
